test/e2e-preferred-host/kcm_preferred_host_kas_test.go

```diff
@@ -1,7 +1,101 @@
+	k8se2eframework "k8s.io/kubernetes/test/e2e/framework"
+
+	configv1client "github.com/openshift/client-go/config/clientset/versioned"
+	operatorv1client "github.com/openshift/client-go/operator/clientset/versioned/typed/operator/v1"
+	"github.com/openshift/cluster-kube-controller-manager-operator/pkg/operator/operatorclient"
+	test "github.com/openshift/cluster-kube-controller-manager-operator/test/library"
+	testlibrary "github.com/openshift/library-go/test/library"
+)
+
+var (
+	// the following parameters specify for how long apis must
+	// stay on the same revision to be considered stable
+	waitForKCMRevisionSuccessThreshold = 3
+	waitForKCMRevisionSuccessInterval  = 1 * time.Minute
+
+	// the following parameters specify max timeout after which
+	// KCMs are considered to not converged
+	waitForKCMRevisionPollInterval = 30 * time.Second
+	waitForKCMRevisionTimeout      = 7 * time.Minute
+)
+
+// TestKCMTalksOverPreferredHostToKAS points the KCM to a non available host
+// and sets unsupported-kube-api-over-localhost flag which changes it to use localhost instead.
+// It then waits for new KCM and tests if creating an SA works.
+func TestKCMTalksOverPreferredHostToKAS(t *testing.T) {
+	// test data
+	kubeConfig, err := test.NewClientConfigForTest()
+	require.NoError(t, err)
+	kubeClient, err := clientset.NewForConfig(kubeConfig)
+	require.NoError(t, err)
+	operatorClientSet, err := operatorv1client.NewForConfig(kubeConfig)
+	require.NoError(t, err)
+	kcmOperator := operatorClientSet.KubeControllerManagers()
+	kcmOperatorConfig, err := kcmOperator.Get(context.TODO(), "cluster", metav1.GetOptions{})
+	require.NoError(t, err)
+
+	openShiftConfigClient, err := configv1client.NewForConfig(kubeConfig)
+	require.NoError(t, err)
+
+	t.Log("getting the current Kube API host")
+	scheme, host := readCurrentKubeAPIHostAndScheme(t, openShiftConfigClient)
+
+	t.Logf("setting the \"unsupported-kube-api-over-localhost\" flag and pointing the current master to %q (non available host)", fmt.Sprintf("%s://%s:1234", scheme, host))
+	data := map[string]map[string][]string{
+		"extendedArguments": {
+			"master":                              []string{fmt.Sprintf("%s://%s:1234", scheme, host)},
+			"unsupported-kube-api-over-localhost": []string{"true"},
+		},
+	}
+	raw, err := json.Marshal(data)
+	require.NoError(t, err)
+	kcmOperatorConfig.Spec.UnsupportedConfigOverrides.Raw = raw
+	_, err = kcmOperator.Update(context.TODO(), kcmOperatorConfig, metav1.UpdateOptions{})
+	require.NoError(t, err)
+	testlibrary.WaitForPodsToStabilizeOnTheSameRevision(t, kubeClient.CoreV1().Pods(operatorclient.TargetNamespace), "kube-controller-manager=true", waitForKCMRevisionSuccessThreshold, waitForKCMRevisionSuccessInterval, waitForKCMRevisionPollInterval, waitForKCMRevisionTimeout)
+
+	// act
+	t.Log("creating a namespace and waiting for a default sa to be populated")
+	testNs, err := k8se2eframework.CreateTestingNS("kcm-preferred-host", kubeClient, nil)
```

Comment:
the `bump (*)` is big because we are pulling `k/k`

Reply:
Can we not bring entire k/k to get just these 2 methods, but just copy them, they're not super hard to write nor sophisticated :cold_sweat:

Reply:
done